### PR TITLE
Snowpack + Typescript conversion (Part 2)

### PR DESCRIPTION
**Pull request by DanilaFe**

### CODE REVIEW

File: src/matrix/storage/idb/stores/RoomMemberStore.ts [hunk 1]
```diff
@@ -15,44 +15,59 @@ See the License for the specific language governing permissions and
...
+// TODO: Move to RoomMember when that's TypeScript.
+export interface MemberData {
+    roomId: string;
+    userId: string;
+    avatarUrl: string;
+    displayName: string;
+    membership: "join" | "leave" | "invite" | "ban";
+}
+
+type MemberStorageEntry = MemberData & { key: string }
```
Comment (bwindels) [inline src/matrix/storage/idb/stores/RoomMemberStore.ts:39]: :smile: nice, basically a shortcut for extending an interface with another.

File: src/matrix/storage/idb/stores/RoomMemberStore.ts [hunk 2]
```diff
@@ -15,44 +15,59 @@ See the License for the specific language governing permissions and
...
+    get(roomId: string, userId: string): Promise<MemberStorageEntry | null> {
         return this._roomMembersStore.get(encodeKey(roomId, userId));
-	}
+    }
 
-	async set(member) {
-        member.key = encodeKey(member.roomId, member.userId);
-        return this._roomMembersStore.put(member);
-	}
+    async set(member: MemberData): Promise<IDBValidKey> {
```
Comment (bwindels) [inline src/matrix/storage/idb/stores/RoomMemberStore.ts]: hmm, ideally we wouldn't leak IDB impl details like `IDBValidKey` from the storage layer. I don't think the returned key is used anywhere, so we can probably avoid the issue altogether by just returning undefined.

File: src/matrix/storage/idb/stores/SessionStore.ts [hunk 3]
```diff
@@ -0,0 +1,48 @@
...
+}
+
+export class SessionStore {
+    private _sessionStore: Store<SessionEntry>
+
+    constructor(sessionStore: Store<SessionEntry>) {
+        this._sessionStore = sessionStore;
+    }
+
+    async get(key: IDBValidKey): Promise<any> {
```
Comment (bwindels) [inline src/matrix/storage/idb/stores/SessionStore.ts]:
```suggestion
    async get(key: string): Promise<any> {
```

File: src/matrix/storage/idb/stores/SessionStore.ts [hunk 4]
```diff
@@ -0,0 +1,48 @@
...
+
+    set(key: string, value: any): Promise<IDBValidKey> {
+        return this._sessionStore.put({key, value});
+    }
+
+    add(key: string, value: any): Promise<IDBValidKey> {
+        return this._sessionStore.add({key, value});
+    }
+
+    remove(key: IDBValidKey): Promise<undefined> {
```
Comment (bwindels) [inline src/matrix/storage/idb/stores/SessionStore.ts]:
```suggestion
    remove(key: string): Promise<undefined> {
```

File: src/matrix/storage/idb/stores/TimelineEventStore.ts [hunk 5]
```diff
@@ -18,22 +18,49 @@ import {EventKey} from "../../../room/timeline/EventKey.js";
...
 
-function decodeEventIdKey(eventIdKey) {
+function decodeEventIdKey(eventIdKey: string): { roomId: string, eventId: string } {
     const [roomId, eventId] = eventIdKey.split("|");
     return {roomId, eventId};
 }
 
 class Range {
-    constructor(IDBKeyRange, only, lower, upper, lowerOpen, upperOpen) {
+    private _IDBKeyRange: any; // TODO what's the appropriate representation here?
```
Comment (bwindels) [inline src/matrix/storage/idb/stores/TimelineEventStore.ts]: [This thing](https://github.com/microsoft/TypeScript/blob/main/lib/lib.dom.d.ts#L9136). Perhaps `typeof IDBKeyRange`?

File: src/matrix/storage/types.ts [hunk 6]
```diff
@@ -14,20 +14,15 @@ See the License for the specific language governing permissions and
...
-    }
-
-    set(invite) {
-        return this._inviteStore.put(invite);
-    }
-
-    remove(roomId) {
-        this._inviteStore.delete(roomId);
-    }
+export interface RoomEvent {
```
Comment (bwindels) [inline src/matrix/storage/types.ts]: We tend to call these timeline events, as state events also appear in a room.

File: src/matrix/storage/idb/stores/TimelineEventStore.ts [hunk 7]
```diff
@@ -231,38 +261,38 @@ export class TimelineEventStore {
...
+     *  @return a promise resolving to undefined if the operation was successful, or a StorageError if not.
      *  @throws {StorageError} ...
      */
-    insert(entry) {
+    insert(entry: StorageEntry): Promise<IDBValidKey> {
         entry.key = encodeKey(entry.roomId, entry.fragmentId, entry.eventIndex);
         entry.eventIdKey = encodeEventIdKey(entry.roomId, entry.event.event_id);
         // TODO: map error? or in idb/store?
-        this._timelineStore.add(entry);
+        return this._timelineStore.add(entry);
```
Comment (bwindels) [inline src/matrix/storage/idb/stores/TimelineEventStore.ts]: So in addition to the earlier comment about not leaking `IDBValidKey` here, the lack of return is also sort off intentional. We don't have any store where keys are automatically created by indexeddb, so we never need the result of a `put` or `add`. Hence, we (I think) never await the promise, but rather make sure the transaction succeeds. The idea is that if there are a lot of `put`/`add`s, this creates a lot of `Promise` instances that are never awaited, so by not exposing it from the storage layer, I wanted to keep the option open of not even calling `reqAsPromise` on modification requests. I forgot the reason I haven't done so already.

Comment (bwindels) [reply src/matrix/storage/idb/stores/TimelineEventStore.ts]: In short, perhaps don't add the returns for now and just make the return type `void`.

Comment (bwindels) [reply src/matrix/storage/idb/stores/TimelineEventStore.ts]: I see now the jsdoc comment contradicts the above, but I just forgot to update it I assume.

**Review by bwindels — COMMENTED** [review]
looks great, just a few small things

**Comment by DanilaFe** [comment]
This should take care of everything you've pointed out here :smile:

**Review by bwindels — APPROVED** [review]

**Comment by bwindels** [comment]
Checks didn't run for some reason on the last commit, but have run them locally and they all pass :heavy_check_mark:

**Pull request merged**